### PR TITLE
add context cancellation checks on merging GetLabel slices

### PR DESCRIPTION
**Pull request by erlan-z**
**What this PR does**:
Add context error check while merging slices for GetLabel operations

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/strings.go [hunk 1]
```diff
@@ -37,17 +38,18 @@ func StringsClone(s string) string {
 
 // MergeSlicesParallel merge sorted slices in parallel
 // using the MergeSortedSlices function
-func MergeSlicesParallel(parallelism int, a ...[]string) []string {
+func MergeSlicesParallel(ctx context.Context, parallelism int, a ...[]string) ([]string, error) {
```
Comment (yeya24) [inline pkg/util/strings.go:41]: Can you add a test case that returns an error due to context check?

Comment (erlan-z) [reply pkg/util/strings.go:41]: I was not sure how to add proper test for context cancellation case. 
 1. I could mock and add some wait time on iteration, and cancel context within this time. But adding wait time on tests? I'm not sure. 
 2. Cancel context before calling the function? but in this one we are not actually testing the real case.
 
There should be better way than the above two. I will add second one, but let me know if we have any other option to properly test it.

Comment (yeya24) [reply pkg/util/strings.go:41]: > Cancel context before calling the function? but in this one we are not actually testing the real case.

I am thinking about before or in the middle of the call. If it is not easy to mock that, I think it is fine to skip.

File: CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +31,7 @@
 * [ENHANCEMENT] Query: Added additional max query length check at Query Frontend and Ruler. Added `-querier.ignore-max-query-length` flag to disable max query length check at Querier. #5808
 * [ENHANCEMENT] Querier: Add context error check when converting Metrics to SeriesSet for GetSeries on distributorQuerier. #5827
 * [ENHANCEMENT] Ruler: Improve GetRules response time by refactoring mutexes and introducing a temporary rules cache in `ruler/manager.go`. #5805
+* [ENHANCEMENT] Querier: Add context error check when merging slices for GetLabel operations. #5837
```
Comment (yeya24) [inline CHANGELOG.md:34]: Should we mention that it is context check when querying Ingesters?

**Review by yeya24 — COMMENTED** [review]
https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L494
Do we want to add context check when we merge all responses?

**Comment by erlan-z** [comment]
> https://github.com/cortexproject/cortex/blob/master/pkg/querier/querier.go#L494 Do we want to add context check when we merge all responses?

Yes, we can add context check there as well, but the code that's used for merge is from thanos. I will create a PR in thanos for that but I think we can merge this PR separately

File: CHANGELOG.md (hunk 2 above)
Comment (erlan-z) [reply CHANGELOG.md]: added

**Comment by yeya24** [comment]
@alanprot Can you also take a look at this PR? Thanks

**Comment by alanprot** [comment]
I think looks good!

**Pull request merged**